AX: WebKit fails to construct text marker ranges that include replaced elements, and never includes AXAttachment in attributed strings, causing various VoiceOver usage issues

twilco opened the pull request:
#### 631a9678ee656600d2e8353511cbbac17786b37b
<pre>
AX: WebKit fails to construct text marker ranges that include replaced elements, and never includes AXAttachment in attributed strings, causing various VoiceOver usage issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=291267">https://bugs.webkit.org/show_bug.cgi?id=291267</a>
<a href="https://rdar.apple.com/148834405">rdar://148834405</a>

Reviewed by Joshua Hoffman.

This commit fixes several bugs introduced by ENABLE(AX_THREAD_TEXT_APIS):

  - We never included the AXAttachment attribute in attributed strings that include attachments, like images. This is
    expected to be an AXUIElement. With this commit, we include this attribute.

  - When posting AXSelectedTextChanged notifications, we must include a valid text marker range, or else VoiceOver won&apos;t
    speak anything. We were failing to create a valid text marker range from VisiblePositions associated with replaced
    elements because AXObjectCache::textMarkerDataForVisiblePosition exited early for anything that wasn&apos;t a RenderText,
    which is a problem for things like images and another replaced elements. Fix this by special-casing replaced elements
    in AXObjectCache::textMarkerDataForVisiblePosition.

  - In AXTextMarkerRange::toAttributedString, we would create our result attributed string based on the text produced
    by the |start| marker. But if the start marker object produced no text, like in the case of a marker with offset 1
    in a replaced element, a nil attributed string would be created. Then subsequent attempts to append more attributed
    strings would fail because they were trying to append to a nil attributed string. Fix this by making an append method
    that detects this scenario.

The two new tests cover all three bugs.

* LayoutTests/accessibility/mac/attributed-string-with-image-expected.txt: Added.
* LayoutTests/accessibility/mac/attributed-string-with-image.html: Added.
* LayoutTests/accessibility/mac/replaced-element-text-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/replaced-element-text-selection.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isReplacedElement const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::createJSStringRef):

Canonical link: <a href="https://commits.webkit.org/293443@main">https://commits.webkit.org/293443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/35838e1afe20b35f397ec399c1a2df81d99d05fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18553 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104055 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49507 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100961 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18848 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27003 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75310 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32428 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101920 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89329 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55671 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14109 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7303 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48885 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/84046 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7377 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106411 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26013 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18961 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84263 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26388 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85526 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83762 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21243 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28423 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6091 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19735 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25966 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31152 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25786 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29106 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27360 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on LayoutTests/accessibility/mac/attributed-string-with-image-expected.txt:
```diff
@@ -0,0 +1,18 @@
+
+TEST COMPLETE
+xyz
+}
```
Looks like this is an extra character (causing this to fail on EWS)

hoffmanjoshua reviewed (approved):
r=me, pending `attributed-string-with-image.html` passing.

twilco committed:
AX: WebKit fails to construct text marker ranges that include replaced elements, and never includes AXAttachment in attributed strings, causing various VoiceOver usage issues
https://bugs.webkit.org/show_bug.cgi?id=291267
rdar://148834405

Reviewed by Joshua Hoffman.

This commit fixes several bugs introduced by ENABLE(AX_THREAD_TEXT_APIS):

  - We never included the AXAttachment attribute in attributed strings that include attachments, like images. This is
    expected to be an AXUIElement. With this commit, we include this attribute.

  - When posting AXSelectedTextChanged notifications, we must include a valid text marker range, or else VoiceOver won't
    speak anything. We were failing to create a valid text marker range from VisiblePositions associated with replaced
    elements because AXObjectCache::textMarkerDataForVisiblePosition exited early for anything that wasn't a RenderText,
    which is a problem for things like images and another replaced elements. Fix this by special-casing replaced elements
    in AXObjectCache::textMarkerDataForVisiblePosition.

  - In AXTextMarkerRange::toAttributedString, we would create our result attributed string based on the text produced
    by the |start| marker. But if the start marker object produced no text, like in the case of a marker with offset 1
    in a replaced element, a nil attributed string would be created. Then subsequent attempts to append more attributed
    strings would fail because they were trying to append to a nil attributed string. Fix this by making an append method
    that detects this scenario.

The two new tests cover all three bugs.

* LayoutTests/accessibility/mac/attributed-string-with-image-expected.txt: Added.
* LayoutTests/accessibility/mac/attributed-string-with-image.html: Added.
* LayoutTests/accessibility/mac/replaced-element-text-selection-expected.txt: Added.
* LayoutTests/accessibility/mac/replaced-element-text-selection.html: Added.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isReplacedElement const):
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::textMarkerDataForVisiblePosition):
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::textRuns):
* Source/WebCore/accessibility/cocoa/AXCoreObjectCocoa.mm:
(WebCore::AXCoreObject::createAttributedString const):
* Source/WebCore/accessibility/cocoa/AXTextMarkerCocoa.mm:
(WebCore::AXTextMarkerRange::toAttributedString const):
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::createJSStringRef):

Canonical link: https://commits.webkit.org/293443@main
